File: testers/duckdb/java/src/java/testdb/DuckDbTestHelper.java
package testdb;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.function.Consumer;
import java.util.function.Function;

public class DuckDbTestHelper {
  private static final String JDBC_URL = "jdbc:duckdb:";
  private static String schemaSQL = null;

  private static synchronized String getSchemaSQL() {
    if (schemaSQL == null) {
      try {
        Path schemaPath = Path.of("sql-init/duckdb/00-schema.sql");
        schemaSQL = Files.readString(schemaPath);
      } catch (IOException e) {
        throw new RuntimeException("Failed to read DuckDB schema", e);
      }
    }
    return schemaSQL;
  }

  private static Connection createConnection() throws SQLException {
    Connection conn = DriverManager.getConnection(JDBC_URL);
    conn.createStatement().execute(getSchemaSQL());
    return conn;
  }

  public static <T> T apply(Function<Connection, T> f) {
    try (Connection conn = createConnection()) {
      conn.setAutoCommit(false);
      try {
        return f.apply(conn);
      } finally {
        conn.rollback();
      }
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }

  public static void run(Consumer<Connection> f) {
    try (Connection conn = createConnection()) {
      conn.setAutoCommit(false);
      try {
        f.accept(conn);
      } finally {
        conn.rollback();
      }
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }
}

File: testers/duckdb/kotlin/src/kotlin/testdb/DuckDbTestHelper.kt
package testdb

import java.nio.file.Files
import java.nio.file.Path
import java.sql.Connection
import java.sql.DriverManager

object DuckDbTestHelper {
    private const val JDBC_URL = "jdbc:duckdb:"
    private val schemaSQL: String by lazy {
        Files.readString(Path.of("sql-init/duckdb/00-schema.sql"))
    }

    private fun createConnection(): Connection {
        val conn = DriverManager.getConnection(JDBC_URL)
        conn.createStatement().execute(schemaSQL)
        return conn
    }

    fun run(f: (Connection) -> Unit) {
        createConnection().use { conn ->
            conn.autoCommit = false
            try {
                f(conn)
            } finally {
                conn.rollback()
            }
        }
    }
}
